### ISSUE TITLE
Text classification accuracy is not stable using cnn or lstm

### ISSUE THREAD

**Issue by Imorton-zd** [fchollet/keras#1610]
When using cnn or lstm to perform text classification muti times, the accuracy of every time is different and vary considerably. For instance, the accuracy of first time is 0.8, then the second time becomes 0.75. In every experiment, the codes are all the same. The main part of my codes as following:

``` python
        tokenizer = Tokenizer(nb_words=max_features)
        X_train = tokenizer.sequences_to_matrix(X_train, mode='binary')
        X_test = tokenizer.sequences_to_matrix(X_test, mode='binary')

        X_train = np.reshape(X_train, (2*n, 1, max_features))    
        X_test = np.reshape(X_test, (1200, 1, max_features))

        print('X_train shape:', X_train.shape)
        print('X_test shape:', X_test.shape)

        print('Build model...')
        model = Sequential()
        model.add(LSTM(lstm_output_dims, input_shape=(1, max_features)))
        model.add(Dense(hidden_dims))
        model.add(Activation('relu'))
        model.add(Dropout(0.5))    
        model.add(Dense(1))
        model.add(Activation('sigmoid'))

        model.compile(loss='binary_crossentropy',optimizer='adam',class_mode="binary")

        print("Train...")
        hist = model.fit(X_train, y_train, nb_epoch=nb_epoch, batch_size=batch_size, verbose=1,shuffle=True, show_accuracy=True,  validation_data=(X_test, y_test))

        print (hist.history)
        print ('train number %s, %s end!\n'%(2*n,d))
        ti =  time.strftime( ISOTIMEFORMAT, time.localtime() )
        f.write('time:%s\n'%ti) 
        f.write('%s bow lstm accuracy: %s %s\n'%(d, max(hist.history['val_acc']), hist.history['val_acc']))
```

**Comment by ymcui** [comment]
Do you mean the training accuracy? or the test time accuracy on a already-trained model?
Because you enabled `shuffle=True` in `model.fit()`, it is possible that the final training accuracy may differ slightly on different runs.

**Comment by Imorton-zd** [comment]
Thanks a lot. @ymcui I mean  the test time accuracy. Setting "shuffle=True" means disorganizing training data but fixed the test data. Or I am wrong?

**Comment by Imorton-zd** [comment]
@ymcui  As far as I am concerned, disorganizing training data but fixed the test data doesn't affect the test time accuracy a lot.

**Comment by pasky** [comment]
How many epochs do you use for training? Does the accuracy converge well before training ends? Did you try with smaller batches?

Another factor that can influence training success is the random initialization.  I'm not completely sure glorot_uniform (the default) is the best idea for RNNs, I'm struggling a bit with a high inter-run variance too.

**Comment by Imorton-zd** [comment]
@pasky nb_epoch is set to 8. I am very sure that the accuracy converges well before training ends. The batches are small enough, or it will train too slowly.
As for the random initialization, I don't know how to operate concretely.

**Comment by audiofeature** [comment]
Add this line just before creating/compiling the model (i.e. before model = Sequential()):

np.random.seed(n)

This sets the random generator of numpy to a repeatable state. n can be any number.

For me, this made all my experiments repeatable with identical results whenever I restart the program.

**Comment by Imorton-zd** [comment]
@audiofeature Thank you very much. In fact, I have enabled shuffle=True in model.fit(), which can achieve the same purpose to your method.

**Comment by tboquet** [comment]
@Imorton-zd setting the random seed and enabling the shuffling of mini-batches is not the same thing. In fact setting `shuffling = True` is directling affecting the variance of your results. Every time you rerun the model, the mini-batches are shuffled differently. If you set the random seed, they will be shuffled the same way and results should be the same accross different runs.

**Comment by Imorton-zd** [comment]
@tboquet So if not setting the random seed and previously the documents are scrambled, the text classification accuracy should be stable?

**Comment by audiofeature** [comment]
Classification accuracies of Neural networks will always be unstable due to their nature involving random weight initialization.

The shuffle=True option helps to shuffle your data to not be biased towards a particular order of the data, which may affect the results.

The only way to have repeatable results (= equal results when you restart the program), is to **use** 
np.random.seed(n)
before initializing the network, as I described.

(This does not mean that the results are more "stable" over several runs, e.g. among cross-validation folds, but that they are "repeatable".) 
(IMHO the only way to get more "stable" results results is to use more data and/or data augmentation)

**Comment by Imorton-zd** [comment]
@audiofeature thank you!  You mean shuffle all data including training and test data? But I want to fix the test data. If only shuffling the taining data, the results can't reappear.

**Comment by xwzhong** [comment]
you can use model.predict_classes() to fix the test data
